src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -0,0 +1,144 @@
+
+        ClientID node_id = ClientID::FromBinary(node_id_string);
+        auto node_info_opt = gcs_client_->Nodes().Get(node_id);
+        RAY_CHECK(node_info_opt)
+            << "Spilling back to a node manager, but no GCS info found for node "
+            << node_id;
+        work.first(nullptr, node_id, node_info_opt->node_manager_address(),
+                   node_info_opt->node_manager_port());
+      }
+    }
+  }
+  return did_schedule;
+}
+
+bool ClusterTaskManager::WaitForTaskArgsRequests(Work work) {
+  Task task = work.second;
+  auto object_ids = task.GetTaskSpecification().GetDependencies();
+  bool can_dispatch = true;
+  if (object_ids.size() > 0) {
+    bool args_ready = fulfills_dependencies_func_(task);
+    if (args_ready) {
+      tasks_to_dispatch_.push_back(work);
+    } else {
+      can_dispatch = false;
+      TaskID task_id = task.GetTaskSpecification().TaskId();
+      waiting_tasks_.try_emplace(task_id, work);
+    }
+  } else {
+    tasks_to_dispatch_.push_back(work);
+  }
+  return can_dispatch;
+}
+
+void ClusterTaskManager::DispatchScheduledTasksToWorkers(WorkerPool &worker_pool) {
+  // Check every task in task_to_dispatch queue to see
+  // whether it can be dispatched and ran. This avoids head-of-line
+  // blocking where a task which cannot be dispatched because
+  // there are not enough available resources blocks other
+  // tasks from being dispatched.
+  for (size_t queue_size = tasks_to_dispatch_.size(); queue_size > 0; queue_size--) {
+    auto task = tasks_to_dispatch_.front();
+    auto reply = task.first;
+    auto spec = task.second.GetTaskSpecification();
+    tasks_to_dispatch_.pop_front();
+
+    std::shared_ptr<Worker> worker = worker_pool.PopWorker(spec);
+    if (!worker) {
+      // No worker available to schedule this task.
+      // Put the task back in the dispatch queue.
+      tasks_to_dispatch_.push_front(task);
+      return;
+    }
+
+    std::shared_ptr<TaskResourceInstances> allocated_instances(
+        new TaskResourceInstances());
+    bool schedulable = cluster_resource_scheduler_->AllocateLocalTaskResources(
+        spec.GetRequiredResources().GetResourceMap(), allocated_instances);
+    if (!schedulable) {
+      // Not enough resources to schedule this task.
+      // Put it back at the end of the dispatch queue.
+      tasks_to_dispatch_.push_back(task);
+      worker_pool.PushWorker(worker);
+      // Try next task in the dispatch queue.
+      continue;
+    }
+
+    worker->SetOwnerAddress(spec.CallerAddress());
+    if (spec.IsActorCreationTask()) {
+      // The actor belongs to this worker now.
+      worker->SetLifetimeAllocatedInstances(allocated_instances);
+    } else {
+      worker->SetAllocatedInstances(allocated_instances);
+    }
+    worker->AssignTaskId(spec.TaskId());
+    worker->AssignJobId(spec.JobId());
+    worker->SetAssignedTask(task.second);
+
+    reply(worker, ClientID::Nil(), "", -1);
```

Comment:
Can you use this syntax to explain what the non-obvious arguments are? `/*argument=*/`